### PR TITLE
NewPanelEditor: Options/FieldConfig API for defaults and common options selection

### PR DESCRIPTION
**Pull request by dprokop**
This PR introduces improves PanelPlugin API:
- adds `useCommonFieldConfig` method - consumer can now select which field config properties panel should support. See docs for example
- adds `devaultValue` to `OptionsEditorItem` interface to allow inline defaults when using options API 
```ts
.addNumberInput({
  id: 'width',
  name: 'Column width',
  description: 'column width (for table)',
  defaultValue: 100, // defaultValue can be now specified inline for panel options
  settings: {
    placeholder: 'auto',
    min: 20,
    max: 300,
  },
})
```
- deprecates `setDafaults` method on `PluginPanel` API
- adds type safety for options API: `id` are now type checked, unfortunately only one level deep as TS does not allow path inferring. Nested paths can still be specified, but they are not typesafe

TODO:
- [x] add defaults support for field config properties

### CODE REVIEW

File: public/app/plugins/panel/table2/module.tsx [hunk 1]
```diff
@@ -35,5 +35,7 @@ export const plugin = new PanelPlugin<Options>(TablePanel)
       id: 'showHeader',
       name: 'Show header',
       description: "To display table's header or not to display",
+      defaultValue: true,
```
Comment (dprokop) [inline public/app/plugins/panel/table2/module.tsx]:
```suggestion
```

File: public/app/plugins/panel/table2/module.tsx [hunk 2]
```diff
@@ -35,5 +35,7 @@ export const plugin = new PanelPlugin<Options>(TablePanel)
...
     });
-  });
+  })
+  .useCommonFieldConfig([StandardFieldConfigProperties.Thresholds]);
```
Comment (dprokop) [inline public/app/plugins/panel/table2/module.tsx]:
```suggestion
  });
```

File: packages/grafana-data/src/panel/PanelPlugin.ts [hunk 3]
```diff
@@ -170,22 +224,37 @@ export class PanelPlugin<TOptions = any> extends GrafanaPlugin<PanelPluginMeta>
...
+   * Allows specyfing which common field config options panel should use
+   *
+   * @example
+   * ```typescript
+   *
+   * import { ShapePanel } from './ShapePanel';
+   *
+   * interface ShapePanelOptions {}
+   *
+   * // when plugin should use only specific common options
```
Comment (ryantxu) [inline packages/grafana-data/src/panel/PanelPlugin.ts]:
```suggestion
   * // when plugin should only display specific common options
```

IIUC, all the common fields are *used*, but this just controls what shows up in the UI

File: packages/grafana-data/src/panel/PanelPlugin.ts [hunk 4]
```diff
@@ -170,22 +224,37 @@ export class PanelPlugin<TOptions = any> extends GrafanaPlugin<PanelPluginMeta>
...
+   * export const plugin = new PanelPlugin<ShapePanelOptions>(ShapePanel)
+   *  .useCommonFieldConfig([StandardFieldConfigProperties.Min, StandardFieldConfigProperties.Max, StandardFieldConfigProperties.Links]);
+   *
+   * // when plugin should use all common options
+   * export const plugin = new PanelPlugin<ShapePanelOptions>(ShapePanel)
+   *  .useCommonFieldConfig();
```
Comment (ryantxu) [inline packages/grafana-data/src/panel/PanelPlugin.ts]: Doing nothing should have all the common fields used?  If you want to hide them could it be:
```
.useCommonFieldConfig([])
```
or maybe `showCommonFieldConfig`?

File: packages/grafana-data/src/types/fieldOverrides.ts [hunk 5]
```diff
@@ -54,16 +54,16 @@ export interface FieldOverrideEditorProps<TValue, TSettings> extends Omit<Standa
...
 }
 
-export interface FieldConfigEditorConfig<TSettings = any> {
-  id: string;
+export interface FieldConfigEditorConfig<TOptions, TSettings = any> {
+  id: (keyof TOptions & string) | string;
   name: string;
   description: string;
   settings?: TSettings;
   shouldApply?: (field: Field) => boolean;
```
Comment (ryantxu) [inline packages/grafana-data/src/types/fieldOverrides.ts]: can you add comment here?  I remember adding this function, but do not remember why!

File: packages/grafana-ui/src/utils/standardEditors.tsx [hunk 6]
```diff
@@ -30,7 +30,7 @@ import { StatsPickerEditor } from '../components/OptionsUI/stats';
  * Returns collection of common field config properties definitions
  */
 export const getStandardFieldConfigs = () => {
```
Comment (ryantxu) [inline packages/grafana-ui/src/utils/standardEditors.tsx]: if we are calling it "commonFieldConfigs" on the panel, we should match that name here.  Either rename this to "common" or the other to "standard"

I'm not sure which is better :)

File: public/app/plugins/panel/table2/module.tsx (hunk 1 above)
Comment (ryantxu) [reply public/app/plugins/panel/table2/module.tsx]: Should the property be "hide header"?  and default to empty?

File: public/app/plugins/panel/table2/module.tsx (hunk 2 above)
Comment (ryantxu) [reply public/app/plugins/panel/table2/module.tsx]: why would table only include thresholds?  I think it would be everything no?

Comment (ryantxu) [reply public/app/plugins/panel/table2/module.tsx]: ahh -- i see... this is on the panel, not the field.  Why would you want threshold on the table, not a cell value?

**Review by ryantxu — APPROVED** [review]
this is looking good.  I added some comments, but in general I like the direction

File: public/app/plugins/panel/table2/module.tsx (hunk 2 above)
Comment (dprokop) [reply public/app/plugins/panel/table2/module.tsx]: @ryantxu this is to be removed. It's only for an example here :)

File: packages/grafana-data/src/panel/PanelPlugin.ts (hunk 3 above)
Comment (dprokop) [reply packages/grafana-data/src/panel/PanelPlugin.ts]: Yes, you are right

File: packages/grafana-data/src/types/fieldOverrides.ts (hunk 5 above)
Comment (dprokop) [reply packages/grafana-data/src/types/fieldOverrides.ts]: yes!

File: packages/grafana-ui/src/utils/standardEditors.tsx (hunk 6 above)
Comment (dprokop) [reply packages/grafana-ui/src/utils/standardEditors.tsx]: your completely right. I think I'll rename common to standard

File: packages/grafana-data/src/panel/PanelPlugin.ts (hunk 4 above)
Comment (torkelo) [reply packages/grafana-data/src/panel/PanelPlugin.ts]: yea, if you want to hide them all I think .useCommonFieldConfig([]) is best, so only need one function to control this

Comment (dprokop) [reply packages/grafana-data/src/panel/PanelPlugin.ts]: @torkelo maybe by default none should be shown unless u use `useCommonFieldConfig`? I think this is most intuitive approach

File: packages/grafana-data/src/types/fieldOverrides.ts (hunk 5 above)
Comment (torkelo) [reply packages/grafana-data/src/types/fieldOverrides.ts]: what is it for, is to check that the value is of right type before passing it to the processor? but can't it be handled inside the processor?

File: packages/grafana-ui/src/utils/standardEditors.tsx (hunk 6 above)
Comment (torkelo) [reply packages/grafana-ui/src/utils/standardEditors.tsx]: yea, was thinking the same thing while reading this that we had 2 names

File: packages/grafana-data/src/panel/PanelPlugin.ts (hunk 4 above)
Comment (torkelo) [reply packages/grafana-data/src/panel/PanelPlugin.ts]: I do not know what default is best (get all standard fields) or optin using useStandardFieldConfig()  with no args.,

File: public/app/plugins/panel/table2/module.tsx (hunk 1 above)
Comment (torkelo) [reply public/app/plugins/panel/table2/module.tsx]: yea was considering this, the Table component property is named noHeader but for a display option not sure what is the most user-friendly / intuitive a default enabled option that you disable to have effect, or a default disable option that you enable to remove the heading. Will ask in #ux

**Review by torkelo — APPROVED** [review]
Great job on this!

**Comment by dprokop** [comment]
@torkelo @ryantxu I've just updated this PR with defaults support for field config properties (both custom - api the same as for panel options, and standard see this doc comment https://github.com/grafana/grafana/pull/23214/files#diff-46ad93593a1e1ee017c96467d9db01ffR271  for more details on the API, it's slightly twisted but I didn't want to introduce another method just for configuring the standard defaults).

Talking about `useStandardFieldConfig` - **by default plugins DON'T use any of the standard configs**. Not sure this is the best approach, but I think this will make using the standard configs more intentional.

File: packages/grafana-data/src/panel/PanelPlugin.ts [hunk 7]
```diff
@@ -42,6 +63,81 @@ export class PanelPlugin<TOptions = any> extends GrafanaPlugin<PanelPluginMeta>
...
+        custom:
+          Object.keys(customPropertiesDefaults).length > 0
+            ? {
+                ...customPropertiesDefaults,
+              }
+            : undefined,
+        ...this._fieldConfigDefaults.defaults,
+      },
+      // TODO: not sure yet what about overrides, if anything
+      overrides: this._fieldConfigDefaults.overrides,
```
Comment (dprokop) [inline packages/grafana-data/src/panel/PanelPlugin.ts]: @ryantxu maybe you have some thoughts on what should we do about default overrides? Personally I don't see why default overrides should be set.

Comment (torkelo) [reply packages/grafana-data/src/panel/PanelPlugin.ts]: I think we can wait with that for now

File: packages/grafana-data/src/panel/PanelPlugin.ts [hunk 8]
```diff
@@ -90,9 +122,11 @@ export class PanelPlugin<TOptions = any, TFieldConfigOptions = any> extends Graf
 
   /**
    * Enables configuration of panel's default field config
+   *
+   * @deprecated setFieldConfigDefaults is deprecated in favor of setCustomFieldOptions
    */
-  setFieldConfigDefaults(defaultConfig: Partial<FieldConfigSource>) {
-    this.fieldConfigDefaults = {
+  setFieldConfigDefaults(defaultConfig: Partial<FieldConfigSource<TFieldConfigOptions>>) {
+    this._fieldConfigDefaults = {
```
Comment (torkelo) [inline packages/grafana-data/src/panel/PanelPlugin.ts]: I think we can remove this, is any plugin using it? was added like 1-2 weeks ago :)

**Comment by torkelo** [comment]
> by default plugins DON'T use any of the standard configs. Not sure this is the best approach, but I think this will make using the standard configs more intentional.

yea I sort of agree.

File: packages/grafana-data/src/panel/PanelPlugin.ts (hunk 8 above)
Comment (dprokop) [reply packages/grafana-data/src/panel/PanelPlugin.ts]: heh, tru dat, think we can remove this

**Comment by dprokop** [comment]
> >by default plugins DON'T use any of the standard configs. Not sure this is the best approach, but I think this will make using the standard configs more intentional.

> yea I sort of agree.

Sort of? :)

**Comment by torkelo** [comment]
I mean I agree :)

**Review by torkelo — APPROVED** [review]
Looks really good now!

**Pull request merged**